Test Utilities/Doubles/Buffer.swift

```diff
@@ -0,0 +1,53 @@
+import XCTest
+@testable import PusherChatkit
+
+public class StubBuffer: DoubleBase, Buffer {
+    
+    private let currentState_toReturn: VersionedState?
+    public private(set) var currentState_actualCallCount: UInt = 0
+    
+    private let filter_toReturn: StateFilter
+    public private(set) var filter_actualCallCount: UInt = 0
+    
+    private let delegate_expectedSetCallCount: UInt
+    public private(set) var delegate_actualSetCallCount: UInt = 0
+    public weak var delegate: BufferDelegate? {
+        didSet {
+            self.delegate_actualSetCallCount += 1
+            
+            guard self.delegate_actualSetCallCount <= self.delegate_expectedSetCallCount else {
+                XCTFail("Unexpected call of `\(#function)` made to \(String(describing: self))", file: self.file, line: self.line)
+                return
+            }
+        }
+    }
+    
+    public init(currentState_toReturn: VersionedState?,
+                filter_toReturn: StateFilter,
+                delegate_expectedSetCallCount: UInt = 0,
+                file: StaticString = #file, line: UInt = #line) {
+        
+        self.currentState_toReturn = currentState_toReturn
+        self.filter_toReturn = filter_toReturn
+        self.delegate_expectedSetCallCount = delegate_expectedSetCallCount
+        
+        super.init(file: file, line: line)
+    }
+    
+    public var currentState: VersionedState? {
+        self.currentState_actualCallCount += 1
+        return self.currentState_toReturn
+    }
+    
+    public var filter: StateFilter {
+        self.filter_actualCallCount += 1
+        return self.filter_toReturn
+    }
+    
+    public func store(_ store: Store, didUpdateState state: VersionedState) {}
+    
+    public func report(_ state: VersionedState) {
+        self.delegate?.buffer(self, didUpdateState: state)
+    }
```

Comment:
I've added `report` methods because:
1. I didn't want to instantiate dependencies in tests just to be able to trigger a delegate method.
2. Not every test is concerned about the dependency that triggered the change, but sometimes you just want to tests the bevaviour when a change occurred (regardless of the trigger of such change).
3. It is more explicit than calling a delegate method.

Reply:
1.  I think this point only applies to `StubJoinedRoomsRepository.report()`.  Thats the only place where the instantiation of anything else is required.  Additionallly, I made a comment elsewhere that mentions that the necessity to instanciate additional entities there is a bit of a code smell.  If we make `JoinedRoomsRepository` a `protocol` as I've suggested then these instanciations go away and I beleive first point is no longer relevant.

2. If you aren't concerned about the dependency that triggered the change then I think you should instead pass dummy sender.  For example `stubStore(DummyStore(), didUpdateState: state)`.  This is more explicit and clearer than hiding it in a convenience method.

3.  I'm confused by this comment.  What could be more explicit than calling the method we are actually trying to test than calling that method directly?

Perhaps we should discuss in person to try and move this forward.  I think its probably worth resolving the `protocol JoinedRoomsRepository` point first though.

Reply:
Since we are postponing the `protocol JoinedRoomsReposity` change for a subsequent PR I think we can postpone the discussion/outcome of this also.  👍